tests/e2e/common/diagnosticcollector.go

```diff
@@ -0,0 +1,564 @@
+	dc.clientset = clientset
+
+	return dc, nil
+}
+
+func (d *DiagnosticCollector) Close() {
+	d.logFile.Close()
+	d.errLogFile.Close()
+}
+
+func (d *DiagnosticCollector) Add(f collectFunc) {
+	d.collectors = append(d.collectors, f)
+}
+
+func (d *DiagnosticCollector) AddExtension(extension string) {
+	switch extension {
+	case NodeFeature:
+		d.Add(d.NodeFeatures)
+	case NodeFeatureRule:
+		d.Add(d.NodeFeatureRules)
+	}
+}
+
+// ParseObjects reads from a comma separated string and sets the DiagnosticCollector objects
+// This is a helper function to set the DiagnosticCollector objects from a comma separated string
+// of objects, for making it easier to set the objects to collect when manually running the
+// diagnostic collector.
+func (d *DiagnosticCollector) ParseObjects(objects string) {
+	for _, obj := range strings.Split(objects, ",") {
+		switch obj {
+		case Nodes:
+			d.Core = append(d.Core, Nodes)
+		case Pods:
+			d.Core = append(d.Core, Pods)
+		case Namespaces:
+			d.Core = append(d.Core, Namespaces)
+		case Deployments:
+			d.Apps = append(d.Apps, Deployments)
+		case DaemonSets:
+			d.Apps = append(d.Apps, DaemonSets)
+		case Jobs:
+			d.Batch = append(d.Batch, Jobs)
+		case NodeFeature:
+			d.Extensions = append(d.Extensions, NodeFeature)
+		case NodeFeatureRule:
+			d.Extensions = append(d.Extensions, NodeFeatureRule)
+		}
+	}
+}
+
+func (d *DiagnosticCollector) RunDiagnostics(ctx context.Context) error {
+	defer d.Close()
+
+	// Collect the core objects
+	for i := range d.Core {
+		switch d.Core[i] {
+		case Nodes:
+			d.Add(d.Nodes)
+		case Pods:
+			d.Add(d.Pods)
+		case Namespaces:
+			d.Add(d.Namespaces)
+		}
+	}
+
+	// Collect the apps objects
+	for i := range d.Apps {
+		switch d.Apps[i] {
+		case Deployments:
+			d.Add(d.Deployments)
+		case DaemonSets:
+			d.Add(d.DaemonSets)
+		}
+	}
+
+	// Collect the batch objects
+	if len(d.Batch) != 0 {
+		// For now we only support jobs
+		d.Add(d.Jobs)
+	}
+
+	// Collect the extensions
+	for i := range d.Extensions {
+		d.AddExtension(d.Extensions[i])
+	}
+
+	if len(d.collectors) == 0 {
+		return fmt.Errorf("no collectors specified")
+	}
+
+	// Run the collectors
+	for _, f := range d.collectors {
+		if err := f(ctx); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Nodes is a collectorFunc that collects node objects
+func (d *DiagnosticCollector) Nodes(ctx context.Context) error {
+	nodes, err := d.clientset.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		if _, lerr := d.errLogFile.WriteString(fmt.Sprintf("Error getting nodes: %v\n", err)); lerr != nil {
+			err = fmt.Errorf("%v+ error writing to stderr log file: %v", err, lerr)
+		}
+		return err
+	}
+
+	nodesFile, err := os.Create(filepath.Join(d.artifactDir, "nodes.yaml"))
+	if err != nil {
+		if _, lerr := d.errLogFile.WriteString(fmt.Sprintf("Error creating nodes.yaml: %v\n", err)); lerr != nil {
+			err = fmt.Errorf("%v+ error writing to stderr log file: %v", err, lerr)
+		}
+		return err
```

Comment:
This is definitely a patter that we're repeating a lot. Can we reduce the boilerplate here?

Reply:
check